SimpleBrowser/Browser.cs

```diff
@@ -25,15 +25,28 @@ namespace SimpleBrowser
     using SimpleBrowser.Parser;
     using SimpleBrowser.Query;
     using System.Security.Cryptography.X509Certificates;
+    using System.Threading.Tasks;
 
-    public class Browser
+    public class Browser : IDisposable

```

Comment:
Why? Browser does not have any unmanaged resources to require implementing IDisposable.

Reply:
It helps the caller who forget to call Close to free up resources. It also fixes a memory leak if there is an Exception before the call to Close is done (or if the caller forgets). If not all frames are cleared there will be pointers to those hanging around. The Dispose method fixes those without the caller needing to know what is going on behind the hood.

Reply:
@mikaelliljedahl SimpleBrowser doesn't have any unmanaged resources, therefore IDisposable is not appropriate. In the case of an Exception, the clean up that is in Close occurs when the instance of Browser is garbage collected.

Reply:
This was to solve this issue:
https://github.com/SimpleBrowserDotNet/SimpleBrowser/issues/121

Even if there are no unmanaged resources, if you have nested references inside a static variable that are never disposed the GC will never be able to do it's job. Since you have this static variable any references inside _allContext will prohibit the GC from freeing up resources.
This is inside Browser class.
private static readonly List<List<Browser>> _allContexts = new List<List<Browser>>();

The dispose method automatically calls the Close method even if exceptions happen. Maybe the Close method needs additional cleanup code to remove references from _allContexts. The important part is to make sure there are no references left from the static list to the disposed object (even if the user missed (eg due to exception) calling Close().

Reply:
According to Microsoft docs:
There are additional reasons for implementing Dispose, such as undoing something that was previously done. For example, freeing memory that was allocated, **_removing an item from a collection_** that was added, signaling the release of a lock that was acquired, and so on

Reply:
In a PR to be implemented shortly, I'm going to fix the "memory leak" permanently by re-working the logging. The user will be able to configure how many pages they would like to navigate back - from 0 (no history) to the amount of memory they have. In my applications, I never use back. I'll probably set mine to 0.